src/compiler/checker.rs

```diff
@@ -2493,37 +2637,175 @@ impl NodeBuilder {
         Some(resulting_node)
     }
 
-    pub fn type_to_type_node_helper(&self, type_: &Type, context: &NodeBuilderContext) -> TypeNode {
-        let synthetic_factory = get_synthetic_factory();
+    pub fn type_to_type_node_helper(
+        &self,
+        type_checker: &TypeChecker,
+        type_: Rc<Type>,
+        context: &NodeBuilderContext,
+    ) -> TypeNode {
         if type_.flags().intersects(TypeFlags::Number) {
             return Into::<KeywordTypeNode>::into(
-                factory.create_keyword_type_node(&synthetic_factory, SyntaxKind::NumberKeyword),
+                factory
+                    .create_keyword_type_node(&self.synthetic_factory, SyntaxKind::NumberKeyword),
             )
             .into();
         }
         if type_.flags().intersects(TypeFlags::Boolean) {
             return Into::<KeywordTypeNode>::into(
-                factory.create_keyword_type_node(&synthetic_factory, SyntaxKind::BooleanKeyword),
+                factory
+                    .create_keyword_type_node(&self.synthetic_factory, SyntaxKind::BooleanKeyword),
             )
             .into();
         }
         if type_.flags().intersects(TypeFlags::BooleanLiteral) {
             return factory
                 .create_literal_type_node(
-                    &synthetic_factory,
+                    &self.synthetic_factory,
                     &*Into::<Rc<Node>>::into(
                         if type_.as_intrinsic_type().intrinsic_name() == "true" {
-                            factory.create_true(&synthetic_factory)
+                            factory.create_true(&self.synthetic_factory)
                         } else {
-                            factory.create_false(&synthetic_factory)
+                            factory.create_false(&self.synthetic_factory)
                         },
                     ),
                 )
                 .into();
         }
+
+        let object_flags = get_object_flags(&*type_);
+
+        if type_.flags().intersects(TypeFlags::TypeParameter)
+            || object_flags.intersects(ObjectFlags::ClassOrInterface)
+        {
+            return if let Some(symbol) = type_.maybe_symbol() {
+                self.symbol_to_type_node(type_checker, symbol, context, SymbolFlags::Type)
+            } else {
+                unimplemented!()
+            };
```

Comment:
Just personal preference, but I'd tend to only use `if let`s if I wasn't going to have an `else`, since with the `else` it's pretty much equivalent to a `match` with the final match being `_`.